### PR TITLE
nixos/victoriametrics: harden systemd unit, add more options.

### PR DESCRIPTION
**Pull request by ryan4yin**
This PR refactors the NixOS Module for VictoriaMetrics, with the structure primarily referencing [the implementation of Prometheus][prometheus.nix] & [vmagent](https://github.com/NixOS/nixpkgs/blob/18536bf04cd71abd345f9579158841376fdd0c5a/nixos/modules/services/monitoring/vmagent.nix#L95).
~~The content of `promTypes.nix` is entirely copied from [prometheus.nix][prometheus.nix] (I'm not quite sure if this part of the configuration can be shared with Prometheus to avoid redundant code)~~.

[prometheus.nix]: https://github.com/NixOS/nixpkgs/blob/nixos-24.05/nixos/modules/services/monitoring/prometheus/default.nix

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

@yorickvP @Shawn8901 @ivan @leona @shawn8901 @ngerstle

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/misc/ids.nix
Comment (ryan4yin) [reply nixos/modules/misc/ids.nix:358]: The previous implementation of this module utilized DynamicUser. However, as a time-series database, the data it holds is valuable and not suited for arbitrary changes to the uid/gid. I believe it's a good idea to assign static uid/gid for it.

If adding it to ids.nix is not appropriate, I can also remove this part and let NixOS dynamically generate the uid/gid instead.

Comment (Ma27) [reply nixos/modules/misc/ids.nix:358]: > However, as a time-series database, the data it holds is valuable and not suited for arbitrary changes to the uid/gid

Not specifying a UID means that the data is still available on the system. You just have to chown it when moving to another system (or back up `/var/lib/nixos` along the way).

The behavior for DynamicUser is similar. With `StateDirectory` a chown is ensured if necessary.

And I think we prefer DynamicUser where possible anyways?

Comment (ryan4yin) [reply nixos/modules/misc/ids.nix]: OK, I’ve reverted the changes to ids.nix and switched back to using DynamicUser.

File: nixos/modules/services/databases/victoriametrics/default.nix
Comment (ryan4yin) [inline nixos/modules/services/databases/victoriametrics/default.nix:59]: In order to be as consistent with Prometheus as possible, I have separated the `port` from the `listenAddress`. This is a breaking change, and I can revert this change if there are any issues.

File: nixos/modules/services/databases/victoriametrics/default.nix
Comment (ngerstle) [inline nixos/modules/services/databases/victoriametrics/default.nix:131]: Prometheus uses a hard-coded user, defined in https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/misc/ids.nix . Can I ask we do the same here?

**Comment by ngerstle** [comment]
> The content of promTypes.nix is entirely copied from [prometheus.nix](https://github.com/NixOS/nixpkgs/blob/nixos-24.05/nixos/modules/services/monitoring/prometheus/default.nix) (I'm not quite sure if this part of the configuration can be shared with Prometheus to avoid redundant code).

I think it would be a good idea to either shared a set of definitions, or remove used options from the copied file..

File: nixos/modules/services/databases/victoriametrics/default.nix
Comment (Ma27) [reply nixos/modules/services/databases/victoriametrics/default.nix:59]: To me this seems a little over the top, I think it's perfectly fine to stick to the victoriametrics way of configuring things here.

**Review by Ma27 — COMMENTED** [review]
> I think it would be a good idea to either shared a set of definitions, or remove used options from the copied file..

Isn't the reason for the equivalent configuration that VM uses the prometheus libraries internally? If so, I'm inclined to agree.

Anyways, I'd like to defer to the VM maintainers, I'm only a user and don't use the scraping part (push agent). cc @leona-ya @yorickvP @Shawn8901 @ivan.

**Comment by Shawn8901** [comment]
> > I think it would be a good idea to either shared a set of definitions, or remove used options from the copied file..
> 
> Isn't the reason for the equivalent configuration that VM uses the prometheus libraries internally? If so, I'm inclined to agree.


i also understood that we are trying to do that, so that we can share.
When we blindly copy the file, i bet they will diverge over time.

Is there actually any option that is currently supported by Prometheus but not by VM, so that there is actually a reason in having two files?

Or we could import the prometheus one and remove unsupported options (if that is possible)?

About the rest of the PR, i did not yet had time to try it out (atm very time limited for coding around and id like to try migrate my config before giving more feedback).

File: nixos/modules/services/databases/victoriametrics/default.nix
Comment (ngerstle) [reply nixos/modules/services/databases/victoriametrics/default.nix:131]: (I'd add that moving to a static/set user becomes important if `vmagent` needs secrets to scrape sources of metrics- passing an agenix encrypted secret to a service using `DynamicUser` isn't feasible, I don't think?)

Comment (Shawn8901) [reply nixos/modules/services/databases/victoriametrics/default.nix:131]: Why not? That works fine if it's passed via systemd tools (Load credentials).
The vmagent module is capable of doing so.

https://github.com/NixOS/nixpkgs/blob/18536bf04cd71abd345f9579158841376fdd0c5a/nixos/modules/services/monitoring/vmagent.nix#L95

**Comment by ryan4yin** [comment]
> Is there actually any option that is currently supported by Prometheus but not by VM, so that there is actually a reason in having two files?

@Shawn8901 I wrote a small script to calculate it, and the results are as follows:

```
supported only by victoriametrics:
{'consulagent_sd_config', 'yandexcloud_sd_config'}
================================================================================
supported only by prometheus:
{'ionos_sd_config',
 'lightsail_sd_config',
 'linode_sd_config',
 'marathon_sd_config',
 'nerve_sd_config',
 'scaleway_sd_config',
 'serverset_sd_config',
 'triton_sd_config',
 'uyuni_sd_config'}
```

Here is the code:

```python3
from pprint import pprint

# https://docs.victoriametrics.com/sd_configs/
victoriametrics_sd = """
azure_sd_configs
consul_sd_configs
consulagent_sd_configs
digitalocean_sd_configs
dns_sd_configs
docker_sd_configs
dockerswarm_sd_configs
ec2_sd_configs
eureka_sd_configs
file_sd_configs
gce_sd_configs
hetzner_sd_configs
http_sd_configs
kubernetes_sd_configs
kuma_sd_configs
nomad_sd_configs
openstack_sd_configs
ovhcloud_sd_configs
puppetdb_sd_configs
static_configs
vultr_sd_configs
yandexcloud_sd_configs
"""

# https://prometheus.io/docs/prometheus/latest/configuration/configuration/
prometheus_sd = """
<azure_sd_config>
<consul_sd_config>
<digitalocean_sd_config>
<docker_sd_config>
<dockerswarm_sd_config>
<dns_sd_config>
<ec2_sd_config>
<openstack_sd_config>
<ovhcloud_sd_config>
<puppetdb_sd_config>
<file_sd_config>
<gce_sd_config>
<hetzner_sd_config>
<http_sd_config>
<ionos_sd_config>
<kubernetes_sd_config>
<kuma_sd_config>
<lightsail_sd_config>
<linode_sd_config>
<marathon_sd_config>
<nerve_sd_config>
<nomad_sd_config>
<serverset_sd_config>
<triton_sd_config>
<eureka_sd_config>
<scaleway_sd_config>
<uyuni_sd_config>
<vultr_sd_config>
<static_config>
"""

victoriametrics_sd = set(victoriametrics_sd.replace(" ", "").replace("configs", "config").splitlines())
prometheus_sd = set(
    prometheus_sd.replace("<", "").replace(">", "").replace(" ", "").splitlines()
)


pprint("supported only by victoriametrics:")
pprint(victoriametrics_sd - prometheus_sd)
pprint("=" * 80)
pprint("supported only by prometheus:")
pprint(prometheus_sd - victoriametrics_sd)
```

Additionally, according to [Unsupported Prometheus config sections - VictoriaMetircs Docs](https://docs.victoriametrics.com/vmagent/#unsupported-prometheus-config-sections), vmagent doesn't support refresh_interval option at service discovery sections.

**Comment by Shawn8901** [comment]
i honestly dislike the extra maintenance burden for having extra  definitions. I like the freeFormFormat that the module currently applies.
It totally lacks the verification (i guess we all agree on that). Both vmagent binary and victoriametrics binary do have a '--dryRun' Flag that verifies the configuration that is given.
Is it okay to just use that? That would also raise all errors of an invalid config file and does not add the extra code to maintain our own set of config definitions, that we need to maintain in sync with VM.

File: nixos/modules/services/databases/victoriametrics.nix
Comment (Shawn8901) [inline nixos/modules/services/databases/victoriametrics.nix:153]: we are missing that option now if i see it correctly.

Comment (Shawn8901) [reply nixos/modules/services/databases/victoriametrics.nix:153]: in case we want to keep the port option change (i dont dislike it), we might want to add a changelog for that, it did at least break my setup as i was using the listenAddress to configure the grafana datasource provisioning, leading to a broken alerting afterwards.

Comment (ryan4yin) [reply nixos/modules/services/databases/victoriametrics.nix]: I have no preference on this option either; I have revert this change & remove these two lines to avoid breaking changes.

**Comment by ryan4yin** [comment]
The PR has been refactored based on the previous review, and it runs well on my Homelab.

My personal configuration (using `BindPaths` and `SupplementaryGroups` to store data in the `/data/apps/victoriametrics` path):

https://github.com/ryan4yin/nix-config/blob/main/hosts/idols-aquamarine/monitoring/victoriametrics.nix

**Review by yorickvP — APPROVED** [review]

**Comment by yorickvP** [comment]
I don't use the scraping part, but I might after this.

**Review by Shawn8901 — APPROVED** [review]
lgtm.
Migrated my setup to the module, tho i am not planning to use the scaping functionality of the single node server, as its non trivial to override the instance label (upstream topic).

**Comment by aanderse** [comment]
hi @ryan4yin :wave: 

i think this PR may have introduced some problems with the module - specifically [here](https://github.com/NixOS/nixpkgs/blob/ac35b104800bff9028425fec3b6e8a41de2bbfff/nixos/modules/services/databases/victoriametrics.nix#L140-L143):

```
        ExecStart = lib.escapeShellArgs (
          startCLIList
          ++ lib.optionals (cfg.prometheusConfig != null) [ "-promscrape.config=${prometheusConfigYml}" ]
        );
```

the `prometheusConfig` option isn't `nullable`, so this doesn't work - and specifically this causes problems for me because i set `promscrape.config` a file that isn't in the nix store (i dynamically generate it)

i think i consider this a regression so a PR to fix this would be justified for a backport to `stable` - do you agree?

i would love to hear your comments on this, thanks kindly!

**Comment by ryan4yin** [comment]
> the `prometheusConfig` option isn't `nullable`, so this doesn't work

Oops, I see that. I will submit a PR to fix it and also add tests to cover this case. Thanks!

**Comment by aanderse** [comment]
thank you so much for offering to take care of this ❤️ 

please ping me on a PR if you can remember to

**Pull request merged**
